app/src/main/java/com/babylon/wallet/android/presentation/dapp/authorized/login/DAppAuthorizedLoginViewModel.kt

```diff
@@ -193,29 +193,42 @@ class DAppAuthorizedLoginViewModel @Inject constructor(
                 }
                 return Result.success(Unit)
             } else {
-                return Result.failure(RadixWalletException.DappRequestException.InvalidPersonaOrAccounts)
+                return Result.failure(InvalidPersonaOrAccounts)
             }
         }
         return Result.success(Unit)
     }
 
+    @Suppress("ReturnCount")
     private suspend fun validateRequestedAddressesForProofOfOwnership(
         requestedPersonaAddress: IdentityAddress?,
         requestedAccountAddresses: List<AccountAddress>?
     ): Boolean {
-        // if requestedPersonaAddress is present, then check if it exists in the allPersonasOnCurrentNetwork
+        // if both requestedPersonaAddress and requestedAccountAddresses are null then return false
+        if (requestedPersonaAddress == null && requestedAccountAddresses == null) return false
+
+        val profile = getProfileUseCase()
+
+        // validate requestedPersonaAddress if present
         val isIdentityValid = requestedPersonaAddress?.let { address ->
-            getProfileUseCase().activePersonasOnCurrentNetwork.any { it.address == address }
-        } ?: true // if requestedPersonaAddress is null, assume it's valid (no need to check)
+            profile.activePersonasOnCurrentNetwork.any { it.address == address }
+        } ?: true // if requestedPersonaAddress is null then it's considered valid
 
-        // if requestedAccountAddresses are present, then check if all of them exist in the existingAccounts
-        val areAccountsValid = requestedAccountAddresses?.let { addresses ->
-            addresses.all { address ->
-                getProfileUseCase().activeAccountsOnCurrentNetwork.any { it.address == address }
-            }
-        } ?: true // if requestedAccountAddresses is null, assume it's valid (no need to check)
+        // if requestedPersonaAddress is not present then return false
+        if (!isIdentityValid) return false
+
+        // if requestedPersonaAddress is valid and requestedAccountAddresses is null then return true
+        if (requestedAccountAddresses == null) return true
+
+        // if requestedPersonaAddress is valid but requestedAccountAddresses is empty then return false
+        if (requestedAccountAddresses.isEmpty()) return false
+
+        // validate requestedAccountAddresses if present and non-empty
+        val areAccountsValid = requestedAccountAddresses.all { address ->
+            profile.activeAccountsOnCurrentNetwork.any { it.address == address }
```

Comment:
Active accounts are queried on every iteration of `all`. Please query all active accounts first and then search if they are valid.

Reply:
another good point 👌🏽